distributed/shuffle/_rechunk.py

```diff
@@ -360,93 +364,88 @@ def __init__(
+        out = defaultdict(list)
+        ndsplits = product(*(axis[i] for axis, i in zip(self.split_axes, partition_id)))
+        for ndsplit in ndsplits:
+            chunk_index, shard_index, ndslice = zip(*ndsplit)
+
+            shard = data[ndslice]
+            # Don't wait until all shards have been transferred over the network
+            # before data can be released
+            if shard.base is not None:
+                shard = shard.copy()
+
+            out[self.worker_for[chunk_index]].append(
+                (chunk_index, (shard_index, shard))
             )
+        out2 = {k: (partition_id, v) for k, v in out.items()}
 
-            for ndsplit in ndsplits:
-                chunk_index, shard_index, ndslice = zip(*ndsplit)
-                out[self.worker_for[chunk_index]].append(
-                    (chunk_index, (shard_index, data[ndslice]))
-                )
-            return {k: (partition_id, pickle.dumps(v)) for k, v in out.items()}
-
-        out = await self.offload(_)
-        await self._write_to_comm(out)
+        await self._write_to_comm(out2)
         return self.run_id
 
     async def _get_output_partition(
         self, partition_id: NDIndex, key: str, **kwargs: Any
     ) -> np.ndarray:
         def _(partition_id: NDIndex) -> np.ndarray:
+            # Quickly read metadata from disk.
+            # This is a bunch of seek()'s interleaved with short reads.
             data = self._read_from_disk(partition_id)
-            return convert_chunk(data)
+            # Copy the memory-mapped buffers on disk onto memory.
+            # This is where we'll spend most time.
+            with self._disk_buffer.time("read"):
+                return convert_chunk(data)
 
         return await self.offload(_, partition_id)
 
-    def deserialize(self, buffer: bytes) -> Any:
-        result = pickle.loads(buffer)
-        return result
-
-    def read(self, path: Path) -> tuple[Any, int]:
-        shards: list[list[tuple[NDIndex, np.ndarray]]] = []
-        with path.open(mode="rb") as f:
-            size = f.seek(0, os.SEEK_END)
-            f.seek(0)
-            while f.tell() < size:
-                shards.append(pickle.load(f))
-        return shards, size
+    def deserialize(self, buffer: Any) -> Any:
+        return buffer
+
+    def read(self, path: Path) -> tuple[list[list[tuple[NDIndex, np.ndarray]]], int]:
+        """Open a memory-mapped file descriptor to disk, read all metadata, and unpickle
+        all arrays. This is a fast sequence of short reads interleaved with seeks.
+        Do not read in memory the actual data; the arrays' buffers will point to the
+        memory-mapped area.
+
+        The file descriptor will be automatically closed by the kernel when all the
+        returned arrays are dereferenced, which will happen after the call to
+        concatenate3.
+        """
+        # Open memory map in read+write mode so that the unpickled numpy arrays can be
+        # writeable
```

Comment:
Why do they need to be writeable?

Reply:
This is a very old argument: https://github.com/dask/distributed/issues/1978

In theory, it sounds like a healthy idea to mark all arrays fed in input to dask tasks as read-only.
In practice, historically third-party libraries (cython) have been problematic with them - so that everything worked fine in unit tests on dask threaded or a single worker, only to fail as soon as the array was transferred between workers. So we decided that dask must honour the read-only flag so that it's the same after serialization as it was before:

## code in distributed
https://github.com/dask/distributed/blob/0ae8110af8688b76e1ddc93c0723027bf0a4bd60/distributed/protocol/numpy.py#L141-L153

## code in zict
https://github.com/dask/zict/blob/7e5dafedcf016a4ac61286badbf1f8da3741d2d3/zict/file.py#L102-L119

I will add a unit test that verifies that the arrays don't accidentally become read-only after shuffle.

Reply:
On second analysis it doesn't matter here because mmap'ed outputs are passed to concatenate3 which forces everything to readable anyway.
I've added a test.

Reply:
Thanks for the explanation!